### PR TITLE
cache registry downloads

### PR DESCRIPTION
**Pull request by tomerd**
motivation: better performance when using registry

cached:
* introduce RegistryDownloadsManager which manages the downloads across the local and cache
* integrate RegistryDownloadsManager into Workspace, which now uses it to download from registry instead of the registry client directly
* update workspace delegate to handle callback from registry downloads, and propogate them to workspace delegates
* improve / generalize the workspace fetching delegate methods such that they are appropriate to both repositories and registry downloads
* deprecate --enable-repository-cache/--disable-repository-cache in favor of --enable-dependency-cache/--disable-dependency-cache
* fix several download and unzip edge cases which came to bear now that a cache is in place
* add and adjust tests

### CODE REVIEW

File: Sources/Commands/Options.swift [hunk 1]
```diff
@@ -152,9 +152,20 @@ public struct SwiftToolOptions: ParsableArguments {
     @Option(help: "Specify the shared security directory")
     var securityPath: AbsolutePath?
 
-    /// Disables repository caching.
-    @Flag(name: .customLong("repository-cache"), inversion: .prefixedEnableDisable, help: "Use a shared cache when fetching repositories")
-    var useRepositoriesCache: Bool = true
+    /// Disables package caching.
+    @Flag(name: .customLong("dependencies-cache"), inversion: .prefixedEnableDisable, help: "Use a shared cache when fetching dependencies")
```
Comment (tomerd) [inline Sources/Commands/Options.swift:156]: not sure if `--disable-dependencies-cache` or `--disable-dependency-cache` is better

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/Commands/Options.swift (hunk 1 above)
Comment (abertelrud) [reply Sources/Commands/Options.swift:156]: Either is correct for an adjectival noun, but it looks as if most of the others use singular (e.g. the one you're replacing as well as things like `--enable-build-manifest-caching`).  So I'd say to go with `--disable-dependency-cache`.

File: Sources/Commands/SwiftTool.swift [hunk 2]
```diff
@@ -74,39 +74,55 @@ private class ToolWorkspaceDelegate: WorkspaceDelegate {
...
+                total: total ?? progress
+            )
+
+            let progress = self.fetchProgress.values.reduce(0) { $0 + $1.progress }
+            let total = self.fetchProgress.values.reduce(0) { $0 + $1.total }
+
+            self.fetchAnimation.update(
+                step: progress > Int.max ? Int.max : Int(progress),
+                total: total > Int.max ? Int.max : Int(total),
+                text: "Fetching \(package)"
```
Comment (abertelrud) [inline Sources/Commands/SwiftTool.swift:121]: For packages that do come from repositories, will there be a way to still see that URL?  Based on how this is called it looks as if this is always the string form of the package identity.

Comment (tomerd) [reply Sources/Commands/SwiftTool.swift:121]: this PR does not change the behavior for repository based dependencies, ie they still pass the url

File: Sources/Commands/SwiftTool.swift [hunk 3]
```diff
@@ -74,39 +74,55 @@ private class ToolWorkspaceDelegate: WorkspaceDelegate {
...
 
-    func fetchingWillBegin(repository: String, fetchDetails: RepositoryManager.FetchDetails?) {
+    func willFetchPackage(package: String, fetchDetails: PackageFetchDetails) {
         queue.async {
-            self.outputStream <<< "Fetching \(repository)"
-            if let fetchDetails = fetchDetails {
-                if fetchDetails.fromCache {
-                    self.outputStream <<< " from cache"
-                }
+            self.outputStream <<< "Fetching \(package)"
```
Comment (abertelrud) [inline Sources/Commands/SwiftTool.swift:79]: For packages that do come from repositories, will there be a way to still see that URL? Based on how this is called it looks as if this is always the string form of the package identity.

Comment (abertelrud) [reply Sources/Commands/SwiftTool.swift:79]: Sorry, moved my comment from one place in the code right as you were replying.  I see, thanks for clarifying.

File: Sources/Commands/SwiftTool.swift (hunk 2 above)
Comment (abertelrud) [reply Sources/Commands/SwiftTool.swift:121]: Sorry, moved the question that this is a reply of just as you were replying.  I see, thanks.

File: Sources/PackageRegistry/RegistryDownloadsManager.swift [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+        version: Version,
+        observabilityScope: ObservabilityScope,
+        delegateQueue: DispatchQueue,
+        callbackQueue: DispatchQueue,
+        completion: @escaping  (Result<AbsolutePath, Error>) -> Void
+    ) {
+        // wrap the callback in the requested queue
+        let completion = { result in callbackQueue.async { completion(result) } }
+        
+        let packageRelativePath: RelativePath!
```
Comment (abertelrud) [inline Sources/PackageRegistry/RegistryDownloadsManager.swift:58]: It's unfortunate that Swift can't reason about this being set in all code paths that don't return.

File: Sources/PackageRegistry/RegistryDownloadsManager.swift [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+        packagePath: AbsolutePath,
+        observabilityScope: ObservabilityScope,
+        delegateQueue: DispatchQueue,
+        callbackQueue: DispatchQueue,
+        completion: @escaping  (Result<FetchDetails, Error>) -> Void
+    ) {
+        if let cachePath = self.cachePath {
+            do {
+                let relativePath = try package.downloadPath(version: version)
+                let cachedPacakgePath = cachePath.appending(relativePath)
```
Comment (abertelrud) [inline Sources/PackageRegistry/RegistryDownloadsManager.swift:145]:
```suggestion
                let cachedPackagePath = cachePath.appending(relativePath)
```

File: Sources/PackageRegistry/RegistryDownloadsManager.swift [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+
+                try self.initializeCacheIfNeeded(cachePath: cachePath)
+                try self.fileSystem.withLock(on: cachedPacakgePath, type: .exclusive) {
+                    // download the package into the cache unless already exists
+                    if try self.fileSystem.validPackageDirectory(cachedPacakgePath) {
+                        // extra validation to defend from racy edge cases?
+                        if self.fileSystem.exists(packagePath) {
```
Comment (abertelrud) [inline Sources/PackageRegistry/RegistryDownloadsManager.swift:152]: What happens if this was partially downloaded?  We had a case recently where the remedy was to remove the preexisting (possibly partial) directory.  Does that apply here?

Comment (abertelrud) [reply Sources/PackageRegistry/RegistryDownloadsManager.swift:152]: Oh never mind, I misread the code.  Marking resolved.

**Review by abertelrud — APPROVED** [review]

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Review by neonichu — APPROVED** [review]
